File: Objc/TapPayTokenPushExample/TapPayTokenPushExample/ViewController.m
//
//  ViewController.m
//  TapPayTokenPushExample
//
//  Created by Cherri_TapPay_LukeCho on 2022/8/9.
//

#import "ViewController.h"
#import "TokenPushViewController.h"
#import "GlobalFunction.h"

#define ACCOUNT @"Test"
#define PASSWORD @"Test"

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UITextField *accountTextField;
@property (weak, nonatomic) IBOutlet UITextField *passwordTextField;
@property (weak, nonatomic) IBOutlet UIButton *loginBtn;
@property (strong, nonatomic) NSArray *queryItems;

@end

@implementation ViewController

#pragma mark - Life Cycle

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(endEdit:)];
    [self.view addGestureRecognizer:tapGesture];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(tokenGet:) name:@"TSP_Push_Token" object:nil];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    dispatch_async(dispatch_get_main_queue(), ^{
        [self setIndicatorHidden:true];
    });
}

#pragma mark - Private Method

- (void)endEdit:(UITapGestureRecognizer *)recognizer {
    [self.view endEditing:true];
}

- (void)tokenGet:(NSNotification *)notification {
    NSArray *queryItems = [notification object];
    _queryItems = queryItems;
    if ([[GlobalFunction valueForKey:@"tspPushToken" fromQueryItems:_queryItems] length] > 0) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self setIndicatorHidden:false];
            [NSTimer scheduledTimerWithTimeInterval:1.0 repeats:false block:^(NSTimer * _Nonnull timer) {
                [self setIndicatorHidden:true];
                [self performSegueWithIdentifier:@"LoginToTokenPush" sender:nil];
            }];
        });
    }
}

- (BOOL)isLoginValid {
    if (_accountTextField.text.length <= 0 ||
        _passwordTextField.text.length <= 0 ||
        ![_accountTextField.text isEqualToString:ACCOUNT] ||
        ![_passwordTextField.text isEqualToString:PASSWORD]) {
        return false;
    }
    return true;
}

#pragma mark - IBAction

- (IBAction)loginPressed:(id)sender {
    if ([self isLoginValid]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self performSegueWithIdentifier:@"LoginToTokenPush" sender:nil];
        });
    }else {
        UIAlertController *controller = [UIAlertController alertControllerWithTitle:@"Error" message:@"Login Failed" preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
        [controller addAction:okAction];
        [self presentViewController:controller animated:false completion:nil];
    }
}

- (IBAction)accountChanged:(id)sender {
    UITextField *textField = (UITextField *)sender;
    [_loginBtn setEnabled:(textField.text.length > 0)];
}

- (IBAction)passwordChanged:(id)sender {
    UITextField *textField = (UITextField *)sender;
    [_loginBtn setEnabled:(textField.text.length > 0)];
}

#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"LoginToTokenPush"]) {
        TokenPushViewController *controller = segue.destinationViewController;
        controller.token = [GlobalFunction valueForKey:@"tspPushToken" fromQueryItems:_queryItems];
        controller.cancelUrl = [GlobalFunction valueForKey:@"cancelUrl" fromQueryItems:_queryItems];
    }
}

@end

File: Swift/TapPayTokenPushExample/TapPayTokenPushExample/ViewController.swift
//
//  ViewController.swift
//  TapPayTokenPushExample
//
//  Created by Cherri_TapPay_LukeCho on 2022/8/11.
//

import UIKit

private let ACCOUNT = "Test"
private let PASSWORD = "Test"

class ViewController: BaseViewController {
    
    @IBOutlet weak var accountTextField: UITextField!
    @IBOutlet weak var passwordTextField: UITextField!
    @IBOutlet weak var loginBtn: UIButton!
    var queryItems : Array<URLQueryItem> = Array.init()
    var token : String = ""
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        NotificationCenter.default.addObserver(self, selector: #selector(tokenGet(notification:)), name: NSNotification.Name.init("TSP_Push_Token"), object: nil)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        setIndicatorHidden(hidden: true)
    }

    @objc func tokenGet(notification : NSNotification) {
        
        queryItems = notification.object as! Array<URLQueryItem>
        token = queryItems.filter({$0.name == "tspPushToken"}).first?.value ?? ""
        if token.count > 0 {
            DispatchQueue.main.async {
                self.setIndicatorHidden(hidden: true)
                Timer.scheduledTimer(withTimeInterval: 1.0, repeats: false) { timer in
                    self.setIndicatorHidden(hidden: false)
                    self.performSegue(withIdentifier: "LoginToTokenPush", sender: nil)
                }
            }
        }
    }
    
    private func isLoginValid() -> Bool {
        if accountTextField.text?.count ?? 0 <= 0 ||
            passwordTextField.text?.count ?? 0 <= 0 ||
            accountTextField.text != ACCOUNT ||
            passwordTextField.text != PASSWORD {
            return false
        }
        return true
    }

    @IBAction func loginBtnPressed(_ sender: Any) {
        if isLoginValid() {
            DispatchQueue.main.async {
                self.performSegue(withIdentifier: "LoginToTokenPush", sender: nil)
            }
        }else {
            let controller = UIAlertController.init(title: "Error", message: "Login Failed", preferredStyle: UIAlertController.Style.alert)
            let okAction = UIAlertAction.init(title: "OK", style: UIAlertAction.Style.default)
            controller.addAction(okAction)
            self.present(controller, animated: false)
        }
    }
    
    @IBAction func accountChanged(_ sender: Any) {
        if let textField = sender as? UITextField {
            loginBtn.isEnabled = (textField.text?.count ?? 0 > 0)
        }
    }
    
    @IBAction func passwordChanged(_ sender: Any) {
        if let textField = sender as? UITextField {
            loginBtn.isEnabled = (textField.text?.count ?? 0 > 0)
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "LoginToTokenPush" {
            if let controller = segue.destination as? TokenPushViewController {
                controller.token = token
                controller.cancelUrl = queryItems.filter({$0.name == "cancelUrl"}).first?.value ?? ""
            }
        }
    }
    
}

